### PR TITLE
feat: add name validation in signup

### PR DESCRIPTION
**Pull request by fcoronelmakingsense**
- Add regular expression to validate name and lastname. use unicode to accept other languages.
- Add non include characters in regular expression, maybe can add more symbols not accepted, i add the more commons.
- I change the message validation because only has one regular expression and tested and is invalid return one message, maybe can generate one message only dont?

examples:

image

image

image

### CODE REVIEW

File: src/components/form-helpers/form-helpers.js [hunk 1]
```diff
@@ -348,6 +349,28 @@ export const InputFieldItem = ({
...
+  ...rest
+}) => (
+  <FieldItem className={concatClasses('field-item', className)} fieldName={fieldName}>
+    <label htmlFor={fieldName}>{label}</label>
+    <Field
+      type="text"
+      name={fieldName}
+      id={fieldName}
+      placeholder={placeholder}
+      validate={combineValidations(createRequiredValidation(required), validateName)}
```
Comment (andresmoschini) [inline src/components/form-helpers/form-helpers.js]: What do you think about adding another flag to _InputFieldItem_ in order to reduce duplicated code?

File: src/i18n/es.js [hunk 2]
```diff
@@ -227,6 +227,7 @@ export default {
     error_invalid_domain_to_register_account: `¡Ouch! Email inválido para crear una cuenta.`,
     error_invalid_email_address: `¡Ouch! El formato del Email es incorrecto`,
     error_invalid_login: `¡Ouch! Hay un error en tu Usuario o Contraseña. Vuelve a intentarlo.`,
+    error_invalid_name: `¡Ouch! Escribe un Nombre válido.`,
```
Comment (andresmoschini) [inline src/i18n/es.js]: This same message applies for both name and lastname fields?

File: src/validations.ts [hunk 3]
```diff
@@ -15,6 +16,21 @@ export function validateEmail(
   return null;
 }
 
+export function validateName(
```
Comment (andresmoschini) [inline src/validations.ts]: Could you add some tests for this function?

File: src/validations.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 const emailRegex = /^([a-zñ\d[\]])(\.?([\wñ&/~\-+\][]+))*@([ñ\w.\-\][]+)\.[\w\-\][.]{2,}(\?)?.*$/i;
+const nameRegex = /^[a-zA-Z][^#&<>[\](\)\"~;$^%{}?]{1,}$/g;
```
Comment (andresmoschini) [inline src/validations.ts]: I am not familiar with quantifiers, could you explain me how this regular expression works?

File: src/validations.ts (hunk 3 above)
Comment (andresmoschini) [reply src/validations.ts]: Please, add a test to verify if the name `Ñoño` works, the same for Chinese characters as you did in the screenshot.

Comment (andresmoschini) [reply src/validations.ts]: And a test to validate if `Andr*s` does not work, should not, right?

Comment (andresmoschini) [reply src/validations.ts]: The same for `+`, `@`, `:`, etc.

File: src/validations.ts (hunk 4 above)
Comment (andresmoschini) [reply src/validations.ts]: Maybe I am wrong, but I am pretty sure that the quantifier is not necessary.

On the other hand, I am not sure that it is ok avoiding unwanted characters with a negation, because there will be always a character that you did not include, for example, an emoji.

File: src/i18n/en.js [hunk 5]
```diff
@@ -225,6 +225,7 @@ export default {
     error_invalid_domain_to_register_account: `Ouch! Invalid Email to create an account.`,
     error_invalid_email_address: `Ouch! Enter a valid Email.`,
     error_invalid_login: `Ouch! There is an error in your Username or Password. Please, try again.`,
+    error_invalid_name: `Ouch! Enter a valid name.`,
```
Comment (cbernat) [inline src/i18n/en.js]: If you use `Nombre` with caps, I think in english should be the same: `Name` right?

File: src/i18n/es.js (hunk 2 above)
Comment (fcoronelmakingsense) [reply src/i18n/es.js]: i put in comments need generic message, maybe can ask Caro for the correct message we can put

File: src/validations.ts (hunk 4 above)
Comment (fcoronelmakingsense) [reply src/validations.ts]: what recommend you??

File: src/validations.ts (hunk 3 above)
Comment (fcoronelmakingsense) [reply src/validations.ts]: ok i test

File: src/i18n/es.js (hunk 2 above)
Comment (andresmoschini) [reply src/i18n/es.js]: Yes, please.

File: src/validations.ts (hunk 4 above)
Comment (andresmoschini) [reply src/validations.ts]: I am not sure, thinking...

Comment (fcoronelmakingsense) [reply src/validations.ts]: i updated the regex because forgot updated

File: src/components/form-helpers/form-helpers.js (hunk 1 above)
Comment (fcoronelmakingsense) [reply src/components/form-helpers/form-helpers.js]: Yes i already push this changes

File: src/i18n/en.js (hunk 5 above)
Comment (fcoronelmakingsense) [reply src/i18n/en.js]: fixed

File: src/components/Signup/Signup.js [hunk 6]
```diff
@@ -7,8 +7,8 @@ import {
   EmailFieldItem,
   FieldGroup,
   FormWithCaptcha,
-  InputFieldItem,
   CheckboxFieldItem,
+  InputFieldItem,
```
Comment (andresmoschini) [inline src/components/Signup/Signup.js]: Why did you moved it?

File: src/components/Signup/Signup.js [hunk 7]
```diff
@@ -152,18 +152,20 @@ const Signup = function({ intl, dependencies: { dopplerLegacyClient, originResol
             <FieldGroup>
               <InputFieldItem
                 autoFocus
+                type="text"
                 className="field-item--50"
                 fieldName={fieldNames.firstname}
                 label={_('signup.label_firstname')}
-                type="text"
```
Comment (andresmoschini) [inline src/components/Signup/Signup.js]: Why did you move it?

File: src/components/Signup/Signup.js [hunk 8]
```diff
@@ -152,18 +152,20 @@ const Signup = function({ intl, dependencies: { dopplerLegacyClient, originResol
...
                 required
+                isValidateName
```
Comment (andresmoschini) [inline src/components/Signup/Signup.js]: What do you think about simply `validateName`?

**Comment by fcoronelmakingsense** [comment]
Update: add unit test

File: src/validations.test.js [hunk 9]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
       expect(result).toEqual({ charLength: true, letter: true, digit: true });
     });
   });
+
+  describe('validateName', () => {
+    it('should not accept an empty name', () => {
```
Comment (andresmoschini) [inline src/validations.test.js]:
```suggestion
    it('should accept an empty name', () => {
```

File: src/validations.test.js [hunk 10]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
...
+      const name = '';
+
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('should not accept an name with one char', () => {
```
Comment (andresmoschini) [inline src/validations.test.js]:
```suggestion
    it('should not accept a name with one char', () => {
```

File: src/validations.test.js [hunk 11]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
...
+      const name = 'a';
+
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toEqual('validation_messages.error_invalid_name');
+    });
+
+    it('should accept an name with two chars', () => {
```
Comment (andresmoschini) [inline src/validations.test.js]:
```suggestion
    it('should accept a name with two chars', () => {
```

File: src/validations.test.js [hunk 12]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
...
+      const name = 'Fe';
+
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('should accept an name with more than one char in chinese language', () => {
```
Comment (andresmoschini) [inline src/validations.test.js]:
```suggestion
    it('should accept a name with more than one char in chinese language', () => {
```

File: src/validations.test.js [hunk 13]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
...
+      const name = '陸軍上校';
+
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('should not accept an name with special character', () => {
```
Comment (andresmoschini) [inline src/validations.test.js]:
```suggestion
    it('should not accept a name with `=` character', () => {
```

File: src/validations.test.js [hunk 14]
```diff
@@ -463,4 +464,73 @@ describe('validations', () => {
...
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toEqual('validation_messages.error_invalid_name');
+    });
+
+    it('should return a custom error', () => {
+      // Arrange
+      const name = 'フェデ+';
```
Comment (andresmoschini) [inline src/validations.test.js]: What do you think about avoiding weird character in order to make the cause of the error more visible?

File: src/validations.test.js (hunk 9 above)
Comment (fcoronelmakingsense) [reply src/validations.test.js]: yes right, sorry

File: src/validations.test.js (hunk 14 above)
Comment (fcoronelmakingsense) [reply src/validations.test.js]: yes you right

**Comment by fcoronelmakingsense** [comment]
@andresmoschini you have more scenarios?

File: src/validations.ts [hunk 15]
```diff
@@ -1,4 +1,5 @@
 const emailRegex = /^([a-zñ\d[\]])(\.?([\wñ&/~\-+\][]+))*@([ñ\w.\-\][]+)\.[\w\-\][.]{2,}(\?)?.*$/i;
+const nameRegex = /^[\u00C0-\u1FFF\u2C00-\uD7FF\w][^#&<>[\](\)\"~;=+@:*$^%{}?]{1,}$/i;
```
Comment (andresmoschini) [inline src/validations.ts]: What do you think about the _negations_?

Maybe this alternative could work: `^[\u00C0-\u1FFF\u2C00-\uD7FF\w][\u00C0-\u1FFF\u2C00-\uD7FF\w'`\-\.\ ]+$`

File: src/components/Signup/Signup.js (hunk 8 above)
Comment (cbernat) [reply src/components/Signup/Signup.js]: Or perhaps `isValidName`?

Comment (andresmoschini) [reply src/components/Signup/Signup.js]: But, it does not represent a state, for that reason, I think that _is_ prefix is wrong.

It is a configuration of the input component.

File: src/validations.ts (hunk 15 above)
Comment (fcoronelmakingsense) [reply src/validations.ts]: update

File: src/components/Signup/Signup.js (hunk 6 above)
Comment (andresmoschini) [reply src/components/Signup/Signup.js]: 😴 💤 💤

File: src/components/Signup/Signup.js (hunk 7 above)
Comment (andresmoschini) [reply src/components/Signup/Signup.js]: 😴 💤 💤

File: src/components/Signup/Signup.js (hunk 8 above)
Comment (andresmoschini) [reply src/components/Signup/Signup.js]: 😴 💤 💤

File: src/i18n/es.js (hunk 2 above)
Comment (andresmoschini) [reply src/i18n/es.js]: Did you get a response from Marketing side?

File: src/validations.test.js [hunk 16]
```diff
@@ -463,4 +464,225 @@ describe('validations', () => {
...
+      // Act
+      const result = validateName(name);
+
+      // Assert
+      expect(result).toEqual('validation_messages.error_invalid_name');
+    });
+
+    it('should not accept a name with special character', () => {
+      // Arrange
+      const name = 'Joh=n';
```
Comment (andresmoschini) [inline src/validations.test.js]: What do you think about moving this scenario to `should not accept a name with special character` (in the array)

**Comment by andresmoschini** [comment]
👍 Please, review the pending comments.

File: src/components/Signup/Signup.js (hunk 6 above)
Comment (fcoronelmakingsense) [reply src/components/Signup/Signup.js]: because create other and when put again move it, i change

File: src/components/Signup/Signup.js (hunk 8 above)
Comment (fcoronelmakingsense) [reply src/components/Signup/Signup.js]: Because is the same name of the function to validate

Comment (fcoronelmakingsense) [reply src/components/Signup/Signup.js]: needValdiateName ?

File: src/i18n/es.js (hunk 2 above)
Comment (fcoronelmakingsense) [reply src/i18n/es.js]: non, caro its busy not respond

File: src/components/Signup/Signup.js (hunk 8 above)
Comment (andresmoschini) [reply src/components/Signup/Signup.js]: _withNameValidation_?

Comment (andresmoschini) [reply src/components/Signup/Signup.js]: _isName_?

File: src/i18n/es.js (hunk 2 above)
Comment (andresmoschini) [reply src/i18n/es.js]: mmm, let me check.

File: src/components/Signup/Signup.js (hunk 7 above)
Comment (fcoronelmakingsense) [reply src/components/Signup/Signup.js]: fixed

File: src/components/Signup/Signup.js (hunk 8 above)
Comment (fcoronelmakingsense) [reply src/components/Signup/Signup.js]: withNameValidation like mroe

**Comment by GusBaamonde** [comment]
Very good @fcoronelmakingsense !!! 👍  :)

File: src/validations.test.js [hunk 17]
```diff
@@ -463,4 +464,215 @@ describe('validations', () => {
       expect(result).toEqual({ charLength: true, letter: true, digit: true });
     });
   });
+
+  describe('validateName', () => {
+    it('should accept an empty name', () => {
```
Comment (cbernat) [inline src/validations.test.js]: Is this ok? Should accept empty names?

Comment (fcoronelmakingsense) [reply src/validations.test.js]: yes, it's another validation. required validation its responsable for that

File: src/i18n/es.js [hunk 18]
```diff
@@ -227,6 +227,9 @@ export default {
     error_invalid_domain_to_register_account: `¡Ouch! Email inválido para crear una cuenta.`,
     error_invalid_email_address: `¡Ouch! El formato del Email es incorrecto`,
     error_invalid_login: `¡Ouch! Hay un error en tu Usuario o Contraseña. Vuelve a intentarlo.`,
+    error_invalid_name: `¡Ouch! Escribe tu nombre usando solo letras.`,
```
Comment (andresmoschini) [inline src/i18n/es.js]: Recordá agregar otro mensaje para apellido.

File: src/i18n/es.js [hunk 19]
```diff
@@ -227,6 +227,9 @@ export default {
...
+    error_min_length: `¡Ouch! Minimo de caracteres invalido.`,
+    error_name_length: `¡Ouch! Escribe al menos dos caracteres.`,
```
Comment (andresmoschini) [inline src/i18n/es.js]: Que te parece renombrar esto como `error_min_length_2`? o algo así, porque creo que se podría usar para otros escenarios, no solo para nombres.

Comment (fcoronelmakingsense) [reply src/i18n/es.js]: tenes razon buen punto, gracias!

File: src/components/form-helpers/form-helpers.js [hunk 20]
```diff
@@ -343,7 +351,15 @@ export const InputFieldItem = ({
       name={fieldName}
       id={fieldName}
       placeholder={placeholder}
-      validate={createRequiredValidation(required)}
+      validate={
+        withNameValidation && minLength
```
Comment (andresmoschini) [inline src/components/form-helpers/form-helpers.js]: Cuidado, si solo se pone uno de los flags, la validación no va a funcionar.

Tal vez se pueda hacer algo así:

```
      validate={ combineValidations(
        required && createRequiredValidation(required),
        minLengt && createMinLengthValidation(minLength),
        withNameValidation && validateName,
      ) }
```

File: src/validations.ts [hunk 21]
```diff
@@ -26,6 +38,18 @@ export function validateRequiredField(
   return null;
 }
 
+export function validateMinLength(
```
Comment (andresmoschini) [inline src/validations.ts]: What do you think about adding some tests for this validation?

Comment (fcoronelmakingsense) [reply src/validations.ts]: yes i need put some

File: src/components/form-helpers/form-helpers.js (hunk 20 above)
Comment (fcoronelmakingsense) [reply src/components/form-helpers/form-helpers.js]: Joya!

File: src/components/form-helpers/form-helpers.js [hunk 22]
```diff
@@ -343,7 +351,11 @@ export const InputFieldItem = ({
       name={fieldName}
       id={fieldName}
       placeholder={placeholder}
-      validate={createRequiredValidation(required)}
+      validate={combineValidations(
+        createRequiredValidation(required),
+        minLength && createMinLengthValidation(minLength),
+        withNameValidation && validateName,
```
Comment (andresmoschini) [inline src/components/form-helpers/form-helpers.js]: Did you test this manually? I mean, for example, what happens when _minLength_ is _falsy_, and _withNameValidation_ is _true_?

Comment (fcoronelmakingsense) [reply src/components/form-helpers/form-helpers.js]: Ready fixed
